shared/src/commonMain/kotlin/io/customer/shared/di/DIGraph.kt

```diff
@@ -0,0 +1,101 @@
+     * We strongly recommend to get instances using methods exposed by Digraph for the purpose, so
+     * it can take responsibility of providing overridden instances where needed.
+     * However, if needed, actual instance should be accessed using the following:
+     * ```
+     * val fileStorage: FileStorage
+     *   get() = override() ?: FileStorage(siteId, context)
+     * ```
+     */
+    inline fun <reified DEP> override(): DEP? = overrides[DEP::class.simpleName] as? DEP
+
+    /**
+     * We prefer to have all of the SDK's singleton instances held in the dependency injection graph.
+     * This makes it easier for automated tests to be able to delete all singletons between each
+     * test function and prevent test flakiness.
+     */
+    val singletons: MutableMap<String, Any> = mutableMapOf()
+
+    /**
+     * In the graph, if you have any dependency that should be a singleton:
+     * ```
+     * val queue: Queue
+     *   get() = getSingletonInstance { QueueImpl(...) }
+     * ```
+     * Note: Don't forget to include the interface in the property declaration:
+     * ```
+     * val foo: InterfaceOfFoo
+     *          ^^^^^^^^^^^^^^
+     * ```
+     * Or you may not be able to successfully mock in test functions.
+     */
+    inline fun <reified INST : Any> getSingletonInstance(crossinline newInstanceCreator: () -> INST): INST {
+        val singletonKey = INST::class.getClassNameOrThrow()
+        val newInstance = { override() ?: newInstanceCreator() }
+        return singletons[singletonKey] as? INST ?: newInstance().also { instance ->
+            singletons[singletonKey] = instance
+        }
+    }
+
+    /**
+     * In the graph, if you have any dependency that should create new instance every time:
+     * ```
+     * val queue: Queue
+     *   get() = getNewInstance { QueueImpl(...) }
+     * ```
+     * Note: Don't forget to include the interface in the property declaration:
+     * ```
+     * val foo: InterfaceOfFoo
+     *          ^^^^^^^^^^^^^^
+     * ```
+     * Or you may not be able to successfully mock in test functions.
+     */
+    inline fun <reified INST : Any> getNewInstance(newInstanceCreator: () -> INST): INST {
+        return override() ?: newInstanceCreator()
+    }
+
+    /**
+     * Call to delete instances held by the graph. This is meant to be called in between automated
+     * tests but can also be called to reset that state of the SDK at runtime.
+     */
+    fun reset() {
+        overrides.clear()
+        singletons.clear()
+    }
+
+    /**
+     * Obtains key using name for the given class.
+     */
+    fun <DEP : Any> KClass<DEP>.getClassNameOrThrow(value: DEP? = null): String {
+        val valueClass = { if (value == null) null else value::class }
+        return simpleName ?: valueClass()?.simpleName
+        ?: throw RuntimeException("Cannot obtain class name for $valueClass")
```

Comment:
Using extension method to extract key and throwing exception as `KClass` name is nullable. Event though it seems working for now, but we might want more appropriate fix before taking it to release. One idea is to get keys in function params which might look like:

Before:
```
val queue: Queue
    get() = getSingletonInstance { QueueImpl(...) }
```
After:
```
val queue: Queue
    get() = getSingletonInstance(key = "Queue") { QueueImpl(...) }
```

Reply:
Docs says for KClass<T>.name/simplename says
```
The simple name of the class as it was declared in the source code, or null if the class has no name (if, for example, it is a class of an anonymous object).
```

On rethinking this, maybe we can do `<T> KClass<T>.java: Class<T>` and then get class name from the java class? if that would be non-nullable?

Reply:
Thanks for adding clarity to the reason @Shahroz16. But unfortunately, `java` class is not available in KMM modules as it should be compatible with both iOS and Android. But from the docs statement, I think we should ideally never get the name as `null` unless we use anonymous objects, which we can easily avoid. I can add comments to the doc just to be safe.

Reply:
> I think we should ideally never get the name as null unless we use anonymous objects, which we can easily avoid.

Because we don't plan on `KClass` being `null`, I don't see a reason to keep this function and instead use `simpleName` in the file. 

If we wanted to feel the most confident that our code will never give us `null` for `KClass`, our automated tests should test that. Possibly through a set of integration tests that use the DiGraph?

Reply:
We are still throwing `RuntimeException` from the method to make sure any bug/unwanted behavior here don't go unnoticed. I used the method to make the reason more clear, but I agree this doesn't look very clean.

But I also don't want to use the `!!` operator as we need a non-nullable key. What do you think of having the following extension
```
val <DEP : Any> KClass<DEP>.simpleClassName: String
    get() {
        return simpleName ?: throw RuntimeException("Cannot obtain class name")
    }
```

And the usage will look like
```
val singletonKey = INST::class.simpleClassName
```

Does it look any better now?